crates/bindings-csharp/BSATN.Runtime/Builtins.cs

```diff
@@ -0,0 +1,134 @@
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using SpacetimeDB.BSATN;
+using SpacetimeDB.Internal;
+
+namespace SpacetimeDB
+{
+    public readonly partial struct Unit
+    {
+        // Custom BSATN that returns an inline empty product type that can be recognised by SpacetimeDB.
+        public readonly struct BSATN : IReadWrite<Unit>
+        {
+            public Unit Read(BinaryReader reader) => default;
+
+            public void Write(BinaryWriter writer, Unit value) { }
+
+            public AlgebraicType GetAlgebraicType(ITypeRegistrar registrar) =>
+                new AlgebraicType.Product(new AggregateElement[] { });
+        }
+    }
+
+    // A helper for special wrappers around byte arrays like Identity and Address.
+    // Makes them equatable, stringifiable, checks length, etc.
+    public abstract record BytesWrapper
+    {
+        protected abstract int SIZE { get; }
+
+        protected readonly byte[] bytes;
+
+        protected BytesWrapper()
+        {
+            bytes = new byte[SIZE];
+        }
+
+        // We can't hide the base class itself, but at least we can hide the constructor.
+        protected BytesWrapper(byte[] bytes)
+        {
+            Debug.Assert(bytes.Length == SIZE);
+            this.bytes = bytes;
```

Comment:
Should these two lines be swapped, in case `SIZE` is defined to return `this.bytes.Length`?

Reply:
Meh. SIZE, as capitalisation sort of implies, should always be implemented as a constant, it's just that you can't have abstract constants in C#.